SVCSE-1595 Setup import of tables from staging FxA databases

akkomar opened the pull request:
Per [SVCSE-1595](https://mozilla-hub.atlassian.net/browse/SVCSE-1595) we'd like to import all the tables from FxA databases to BigQuery. There are 37 tables per environment (stage and production). If we follow the current approach we'd end up having all those tables in `accounts_backend_external` dataset, and their views in `accounts_backend` together with telemetry data. This layout will make finding data hard.

To better organize this data I propose to:
* introduce two new external datasets: `accounts_db_nonprod_external` and `accounts_db_external` to replace `accounts_backend_external` and store data imported from FxA databases
* introduce two datasets to store user-facing views to these tables: `accounts_db_nonprod` and `accounts_db`
* remove `accounts_backend_external` dataset
* remove previously created user-facing views: `accounts_backend.accounts` and `accounts_backend.emails`

@kik @srose this is still work in progress, but I'd like to hear your opinions on the proposed approach. Currently this PR partially implements first step from above plan: adds `accounts_db_nonprod_external` and all the tables we'd want to import from stage.


Checklist for reviewer:

- [ ] Commits should reference a bug or github issue, if relevant (if a bug is referenced, the pull request should include the bug number in the title).
- [ ] If the PR comes from a fork, trigger integration CI tests by running the [Push to upstream workflow](https://github.com/mozilla/bigquery-etl/actions/workflows/push-to-upstream.yml) and provide the `<username>:<branch>` of the fork as parameter. The parameter will also show up
in the logs of the `manual-trigger-required-for-fork` CI task together with more detailed instructions.
- [ ] If adding a new field to a query, ensure that the schema and dependent downstream schemas have been updated.
- [ ] When adding a new derived dataset, ensure that data is not available already (fully or partially) and recommend extending an existing dataset in favor of creating new ones. Data can be available in the [bigquery-etl repository](https://github.com/mozilla/bigquery-etl), [looker-hub](https://github.com/mozilla/looker-hub) or in [looker-spoke-default](https://github.com/mozilla/looker-spoke-default/tree/e1315853507fc1ac6e78d252d53dc8df5f5f322b).

For modifications to schemas in restricted namespaces (see [`CODEOWNERS`](https://github.com/mozilla/bigquery-etl/blob/main/CODEOWNERS)):
- [ ] Follow the [change control procedure](https://docs.google.com/document/d/1TTJi4ht7NuzX6BPG_KTr6omaZg70cEpxe9xlpfnHj9k/edit#heading=h.ttegrcfy18ck)


[SVCSE-1595]: https://mozilla-hub.atlassian.net/browse/SVCSE-1595?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/DENG-1992)

sean-rose commented on sql/moz-fx-data-shared-prod/accounts_db_nonprod_external/fxa_accounts_v1/schema.yaml:
```diff
@@ -0,0 +1,44 @@
+- name: emailVerified
+  type: BOOLEAN
+  mode: NULLABLE
+- name: verifierVersion
+  type: BOOLEAN
```
Not all tinyints are used for boolean values.
```suggestion
- name: verifierVersion
  type: INTEGER
```

sean-rose commented on sql/moz-fx-data-shared-prod/accounts_db_nonprod_external/fxa_email_bounces_v1/schema.yaml:
```diff
@@ -0,0 +1,19 @@
+- name: email
+  type: STRING
+  mode: NULLABLE
+- name: bounceType
+  type: BOOLEAN
+  mode: NULLABLE
+- name: bounceSubType
+  type: BOOLEAN
+  mode: NULLABLE
+- name: createdAt
+  type: TIMESTAMP
+  mode: NULLABLE
+- name: emailTypeId
+  type: BOOLEAN
+  mode: NULLABLE
```
```suggestion
- name: bounceType
  type: INTEGER
  mode: NULLABLE
- name: bounceSubType
  type: INTEGER
  mode: NULLABLE
- name: createdAt
  type: TIMESTAMP
  mode: NULLABLE
- name: emailTypeId
  type: INTEGER
  mode: NULLABLE
```

sean-rose commented on sql/moz-fx-data-shared-prod/accounts_db_nonprod_external/fxa_linked_accounts_v1/schema.yaml:
```diff
@@ -0,0 +1,17 @@
+- name: id
+  type: STRING
+  mode: NULLABLE
+- name: providerId
+  type: BOOLEAN
```
```suggestion
- name: providerId
  type: INTEGER
```

sean-rose commented on sql/moz-fx-data-shared-prod/accounts_db_nonprod_external/fxa_account_customers_v1/schema.yaml:
```diff
@@ -0,0 +1,14 @@
+  type: STRING
+  mode: NULLABLE
+  description: Account ID in hexadecimal format.
+- name: stripeCustomerId
```
I wonder if we should go to the hassle of making the BigQuery columns snake_case.  We're essentially already doing so for the table names, and if any of these tables will be made available in Looker that would impact how the column names would be displayed there (though I suppose we could also implement LookML generator logic to create snake_case dimensions for camelCase columns).

akkomar replied:
I was expecting this data would be exposed in Looker via aggregates and raw tables would rather be used for exploratory analyses. One argument for keeping columns unchanged I can think of is that it makes it a bit easier to find them in FxA codebase. I'm open to discussion though.

sean-rose replied:
I'm fine punting on this for now.  No need to go to all that extra work if there isn't a concrete use case/benefit.

sean-rose commented:
(note that because my review had more than 10 comments GitHub auto-hides some comments in the middle, which is easy to miss)

akkomar commented on sql/moz-fx-data-shared-prod/accounts_db_nonprod_external/fxa_account_reset_tokens_v1/query.sql:
```diff
@@ -0,0 +1,13 @@
+FROM
+  EXTERNAL_QUERY(
+    "moz-fx-fxa-nonprod.us.fxa-rds-nonprod-stage-fxa",
+    """SELECT
+         uid,
+         createdAt
+       FROM
+         fxa.accountResetTokens
```
Yeah, it's unfortunate choice of PK. Might make the table useless in BQ.

akkomar commented on sql/moz-fx-data-shared-prod/accounts_db_nonprod_external/metadata.yaml:
```diff
@@ -0,0 +1,14 @@
+  Data extracted from the nonprod (stage) FxA backend services databases.
+  See https://mozilla.github.io/ecosystem-platform/reference/database-structure#database-fxa for more information.
+
+  Access to this dataset is restricted to accounts-confidential workgroup because some tables here contain sensitive data.
+dataset_base_acl: restricted
+user_facing: false
+labels: {}
+workgroup_access:
+- role: roles/bigquery.dataViewer
+  members:
+  - workgroup:accounts-confidential
```
Good point, but per discussion in ]the ticket](https://mozilla-hub.atlassian.net/browse/SVCSE-1595?focusedCommentId=797812) let's keep this restricted too.

akkomar commented:
@sean-rose thanks for the comments. I've incorporated your suggestions in the PR, fixed some additional casting issues, and tested these queries locally.
This PR now configures import of all tables from staging environment and is ready for a final review.

sean-rose commented on sql/moz-fx-data-shared-prod/accounts_db_nonprod_external/metadata.yaml:
This file should be named `dataset_metadata.yaml`, not `metadata.yaml`.

sean-rose reviewed (approved):
r+wc
